Hydration fix (#523)
kirill-konshin/next-redux-wrapper: README.md

      ...(await App.getInitialProps(context)).pageProps,
      // Some custom thing for all pages
      pathname: ctx.pathname,
    },
  };
});

export default MyApp;
```

Then all pages can be [instrumented the same way](#pagegetinitialprops).

:warning: You can use `getServerSideProps` or `getStaticProps` at page level while having `App.getInitialProps`, this scenario is supported, but I highly don't recommend to do this. In this case you should expect 2 `HYDRATE` actions being dispatched:

1. With state after `App.getInitialProps`
2. With state after `getServerSideProps` or `getStaticProps`

If you use `getServerSideProps` at page level `getServerSideProps` will be executed **after** `App.getInitialProps`, so first dispatched `HYDRATE` will be incomplete, and second will have state from both.

If you use `getStaticProps` at page level then `getStaticProps` will be executed at **compile time** and store there will **NOT** have state from `App.getInitialProps` because they are executed in different contexts and state cannot be shared, so both `HYDRATE`s will have portions of state along with some default values (if any).

# State reconciliation during hydration

> Each time when pages that have `getStaticProps` or `getServerSideProps` are opened by user the `HYDRATE` action will be dispatched. This may happen during initial page load and during regular page navigation. The `payload` of this action will contain the `state` at the moment of static generation or server side rendering, so your reducer must merge it with existing client state properly.

Reviewer: ```suggestion
Each time the user opens a page containing the `useHydration` hook, the `HYDRATE` action will be dispatched. This may happen during initial page load and during regular page navigation. The `payload` of this action will contain the `state` at the moment of static generation or server side rendering, so your reducer must merge it with existing client state properly.
```

- Phase 1: `getInitialProps`/`getStaticProps`/`getServerSideProps`

>   - The wrapper creates a server-side store (using `makeStore`) with an empty initial state. In doing so it also provides the `Request` and `Response` objects as options to `makeStore`.

Reviewer: ```suggestion
  - The wrapper creates a server-side store using `makeStore`.
```
We don't provide an initial store state, so the initial state only depends on user code. Due to GSP, req and res are not necessarily provided as well.

  - The wrapper calls the `_app`'s `getInitialProps` (if any) function and passes the previously created store.

>   - Next.js takes the props returned from the `_app`'s `getInitialProps` method, along with the store's state.

Reviewer: ```suggestion
  - Next.js takes the props returned from the `_app`'s `getInitialProps` method, as well as a serialized list of actions that were dispatched.
```

  - The wrapper calls the Page's `getXXXProps` function and passes the previously created store.

>   - Next.js takes the props returned from the Page's `getXXXProps` method, along with the store's state.

Reviewer: ```suggestion
  - Next.js takes the props returned from the Page's `getXXXProps` method, as well as a serialized list of actions that were dispatched.
```

- Phase 2: SSR

  - The wrapper creates a new store using `makeStore`
  - The wrapper replays all actions dispatched on Phase 1
  - Resulting HTML is emitted to browser

>   - **Connected components may alter the store's state, but the modified state will not be transferred to the client.**

Reviewer: ```suggestion
  - **Connected components may dispatch actions in this phase too, but they will not be replayed on the client.**
```

- Phase 3: Client

  - The wrapper creates a new store

>   - The wrapper replays all actions dispatched on Phase 1

Reviewer: ```suggestion
  - The wrapper replays all actions dispatched in Phase 1
```

- Phase 4: Soft client navigation (without reload)
  - The wrapper reuses same client store

>   - The wrapper idles if no `getServerSideProps` or `getStaticProps` were used on the page, otherwise the wrapper replays all actions dispatched on new Phase 1 of the new page

Reviewer: ```suggestion
  - The wrapper idles if no `getServerSideProps` or `getStaticProps` were used on the page, otherwise the wrapper replays all actions dispatched in Phase 1 of the new page
```

export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;
```

> ## Server and Client state separation

Reviewer: I don't think we need this section anymore?

Author: Still needed, because I still recommend not to mix them in order not to write too smart reducers.
